### ISSUE TITLE
Data arguments are not handled correctly during chunkrun

### ISSUE THREAD

**Issue by bentsherman** [SystemsGenetics/ACE#65]
Here's the script I'm trying:
```
SIZE=2

kinc settings set opencl none

for (( i = 0; i < $SIZE; i++ )); do
	kinc chunkrun $i $SIZE similarity --input data/Yeast.emx --ccm data/Yeast-chunk.ccm --cmx data/Yeast-chunk.cmx &
done

wait
```

And the output:
```
Similarity::Input::set(index=0)
Similarity::Input::set(index=0)

../../src/core/similarity.cpp:253
virtual void Similarity::initialize()
INVALID ARGUMENT
Did not get valid input and/or output arguments.

../../src/core/similarity.cpp:253
virtual void Similarity::initialize()
INVALID ARGUMENT
Did not get valid input and/or output arguments.
```

I added a print statement for every time the input analytic is called with a data argument. It looks like the `--input` argument is passed but `--ccm` and `--cmx` are not.

**Comment by 4ctrl-alt-del** [comment]
I cannot replicate your error in ACE using the example program. I run the following bash script:

~~~
#!/bin/bash

SIZE=2

aceex settings set opencl none

for (( i = 0; i < $SIZE; i++ )); do
	aceex chunkrun $i $SIZE transform --in a.num --out b.num --type addition --amount 5 &
done

wait

aceex merge $SIZE transform --in a.num --out b.num --type addition --amount 5
~~~

It works without any error. All arguments are captured. There must be something wrong in your KINC analytic.

**Comment by 4ctrl-alt-del** [comment]
In an ironic twist trying to replicate your error made me find another unrelated error in the chunk run analytic that caused a bus error due to a corrupted qt object pointer. :)

**Comment by bentsherman** [comment]
I found the source of the problem, sort of. The ACE example doesn't fail because it doesn't check for a valid output argument during `initialize()`. In a normal run this can lead to a segfault if the output argument is not provided (refer to #58). In a chunk run, the segfault never happens because the output data object is never actually written until merge.

However, in KINC similarity, I need the output arguments in `initialize()` because the data objects themselves need to be initialized. That being said, the only other place where I use them is in `process()`, so I should be able to move the initialization code to `process()` in order to make chunk run work.

I think this might be a design issue with chunk run though. From the end user perspective, it's intuitive to assume that output data objects will be available during `initialize()`. By the same token, the idea of chunkrun is that it should work for any analytic out of the box by simply separating the execution and processing of work blocks, and so the user shouldn't have to implement an analytic with the details of chunk run in mind.

**Comment by bentsherman** [comment]
So I think we probably need to have the output data objects available during `initialize()`, because otherwise we can't validate them and prevent segfaults, and it's an intuitive assumption since we have all other input arguments. But the solution may be to have two initialize methods, for example, `initializeInput()` / `initializeOutput()` or `initialize()` / `setupOutputs()`. The latter is probably better since an analytic might have initialization unrelated to input arguments. But it's hard to say because Similarity is the only KINC analytic that generates work blocks, for which chunkrun is applicable.

**Comment by bentsherman** [comment]
Or I suppose we could add something like `isChunk()` to the analytic class, so that the analytic can detect whether or not it's in a chunk run and handle output data objects accordingly.

**Comment by 4ctrl-alt-del** [comment]
I took your suggestion and added the virtual function initializeOutputs() to analytics. This is only called on analytics that will be used for output. So it is not called for chunk or mpi slave analytics. Fixed in commit 693a80600dd178abe210fd9d35aaebee1f19a21a. Let me know if this resolves your issue.

**Comment by bentsherman** [comment]
The fix works, resolved in the latest commit of KINC.
